Adding additional content to new 'variables' page (#581)
okteto/docs: src/content/manifest/okteto-variables.mdx

---
title: Okteto Variables
description: Okteto Variables allows you to save application configuration in Okteto and automatically inject them during deployment time
sidebar_label: Variables
id: okteto-variables
---

import Image from "@theme/Image";

Application configuration should be passed at deployment time, not hardcoded in your code.

Okteto variables help you save application configuration in Okteto and automatically inject it during deployment time.

There are a few different ways to leverage variables. In short:

- You can define variables manually in the manifest or in the Okteto UI
- There are some default variables built in to the Okteto platform
- All variables used in the Okteto manifest are automatically expanded

> - Variables referenced in the `deploy` and `destroy` sections are automatically injected into those environments

Author: @pchico83 is this an accurate representation? It felt right to me but wondering if I've captured your suggestion here

Reviewer: Maybe something like `All variables are injected as runtime environment variables in the execution of your "deploy" and "destroy" scripts`?

Author: Done (locally) 👍 Will push in next commit

We will cover each of these behaviors in more detail below.

## User-defined variables

Okteto supports manual definition of variables, offering significant flexibility in configuration. These user-defined variables are separate from Okteto's default, built-in variables. We recommend creating variables in the following order of preference:

- Local variables
- Deploy time variables
- Catalog variables
- User variables (in settings)
- Admin variables

> > Note: Variables defined using methods lower in this list (e.g. admin variables) will override mentions of the same variable defined by methods higher in this list (e.g. deploy time variables)

Author: @pchico83 I added this because I assume this is the case. How should I correct this?

Reviewer: Preference is in reverse order. For example, a deploy time variable overrides an admin variable

Author: Thank you! Will push change in next commit 👍

  />
</p>

## Configuring dynamic endpoints

> You can create dynamic endpoints for external resources within your Okteto environment using `$OKTETO_ENV`. You can define `$variables`, assign them to an `OKTETO_ENV_VAR`, and make those ENV VARs available in other sections (e.g. `deploy`, `destroy`) of the Oketo manifest to generate dynamic endpoints for external resources. 

Reviewer: ```suggestion
You can create dynamic endpoints for external resources within your Okteto environment using `$OKTETO_ENV`. You can define `$variables`, assign them to `OKTETO_ENV`, and make those variables available in other sections (e.g. `deploy`, `destroy`) of the Okteto manifest or to generate dynamic endpoints for external resources. 
```

Author: Making this change locally as part of a batch update. Will push in next commit 👍

> The example below demonstrates adding variables to `$OKTETO_ENV` in the `deploy` section and then access those variables in a sample `destroy` section:

Author: @pchico83 I re-wrote this based on your feedback. WDYT?

```yaml
deploy:
  hashicorp/terraform:1.4
  commands:
  - name: Create AWS infrastructure
    command: |
      set -e
      resourceName="${OKTETO_NAMESPACE}-oktacoshop"
      region=$AWS_REGION

      terraform apply -input=false -var "bucket_name=$resourceName" -var "queue_name=$resourceName" -var "region=$region" -auto-approve
      s3Dashboard="https://s3.console.aws.amazon.com/s3/buckets/${resourceName}"

      # ... shortened for brevity
    {
      echo "OKTETO_EXTERNAL_S3_ENDPOINTS_BUCKET_URL=$s3Dashboard"
      echo "S3_BUCKET_NAME=$resourceName"
    } >> "$OKTETO_ENV"

destroy:
  image: hashicorp/terraform:1.4
  commands:
  - name: Delete the AWS infrastructure
    command: |
      set -e
      resourceName="${OKTETO_NAMESPACE}-oktacoshop"
      region=$AWS_REGION
      
      # omitted for brevity 

      terraform apply -input=false -var "bucket_name=$resourceName" -var "queue_name=$resourceName" -var "region=$region" -auto-approve --destroy

> ```

Reviewer: Awesome, answered [here](https://okteto.slack.com/archives/C043X0H48BZ/p1704869018189579?thread_ts=1704748882.478229&cid=C043X0H48BZ)

### **Default Environment Variables**

> These are automatically populated and can be referenced in your Okteto manifest’s `deploy`/`destroy` commands:

Author: @pchico83 I had suggested we remove the word "commands" here to avoid confusion since I understood your previous comment to mean variables can be used anywhere in the manifest YAML and not solely, specifically within the `commands:` config

Reviewer: There is a slightly difference. All variables in the okteto manifest are expanded. But in th deploy/destroy commands they are also injected in the execution runtime. That means, my deploy section could by:
```
deploy:
   - make deploy
```
And I have a `Makefile` where `make deploy` is defined to access okteto variables and that would work.
In other words, `make deploy` doesn't have references to these variables, but its internal implementation does, and that works because we inject all these variables in the runtime environment of the deploy/destroy commands

Author: What about this?

> These are automatically expanded and can be referenced in your Okteto manifest’s `deploy`/`destroy` sections either directly (within the manifest) or in external files (e.g. a `Makefile`):

Reference variables defined by your dependencies using the following pattern:

- **`OKTETO_DEPENDENCY_<dependency>_VARIABLE_<variable>`**: The value of a variable of a dependency.
    
For example, if you have a dependency in your Okteto Manifest:

```yaml
dependencies:
  database:
      repository: https://github.com/okteto/my-database
      variables:
      USERNAME: okteto
      wait: true
```

And the repository has this Okteto Manifest:

```yaml
deploy:
  - echo "PASSWORD=1234" >> $OKTETO_ENV
```

Reference the variables in your Okteto Manifest deploy section like this:

```yaml
dependencies:
  database:
    repository: https://github.com/okteto/my-database
    variables:
    USERNAME: okteto
    wait: true
deploy:
  - echo ${OKTETO_DEPENDENCY_DATABASE_VARIABLE_USERNAME}
  - echo ${OKTETO_DEPENDENCY_DATABASE_VARIABLE_PASSWORD}
```
    
> Note: USERNAME has a static value, and PASSWORD is dynamically generated.

## Feature flag variables

Okteto provides a list of variables that serve as feature flags for some of the features we release in our product. Below is a list of those variables that you can use to leverage these newer and experimental features:

> - `OKTETO_SMART_BUILDS_ENABLED`: enables the "smart builds" feature that uses a hash of the build context to determine if an image needs to be rebuilt or can be pulled from the registry. This can accelerate build times.

Author: @pchico83 Are there any other feature flag variables to include here?

Author: I'm supportive of documenting them now quickly and concisely. A follow-on iteration could expand further if we need to.

Is there an internal document or other resource to help me understand what each does so I can populate this content?

Reviewer: @jLopezbarb could you help to document these variables?

Author: Resolving this to merge the PR since it's approved and all other notes are addressed.

Will submit a follow-up PR once we have this list with details 👍

> ## Accessing Okteto Variables from your Okteto Pipeline

Reviewer 2: The content of this is outdated, I just sent a PR to update the tutorial repo.

```suggestion
## Accessing Okteto Variables at deploy time
```

I also updated the paragraph below via a commit to this branch (I couldn't do it via the review suggestion).

Author: Thank you! I've made the change locally and will push in next commit
